src/widget/wtrackmenu.cpp

```diff
@@ -0,0 +1,1548 @@
+        showTrackInfo(prevRow);
+        if (!m_pTagFetcher.isNull()) {
+            showDlgTagFetcher(prevRow);
+        }
+    }
+}
+
+void WTrackMenu::showTrackInfo(QModelIndex index) {
+    DEBUG_ASSERT(m_pTrackModel);
+    TrackModel* trackModel = m_pTrackModel;
+
+    if (!trackModel) {
+        return;
+    }
+
+    if (m_pTrackInfo.isNull()) {
+        // Give a NULL parent because otherwise it inherits our style which can
+        // make it unreadable. Bug #673411
+        m_pTrackInfo.reset(new DlgTrackInfo(m_pConfig, nullptr));
+
+        connect(m_pTrackInfo.data(), SIGNAL(next()),
+                this, SLOT(slotNextTrackInfo()));
+        connect(m_pTrackInfo.data(), SIGNAL(previous()),
+                this, SLOT(slotPrevTrackInfo()));
+        connect(m_pTrackInfo.data(), SIGNAL(showTagFetcher(TrackPointer)),
+                this, SLOT(slotShowTrackInTagFetcher(TrackPointer)));
+        connect(m_pTrackInfo.data(), SIGNAL(finished(int)),
+                this, SLOT(slotTrackInfoClosed()));
+    }
+    TrackPointer pTrack = trackModel->getTrack(index);
+    m_pTrackInfo->loadTrack(pTrack); // NULL is fine.
+    currentTrackInfoIndex = index;
+    m_pTrackInfo->show();
+}
+
+void WTrackMenu::slotNextDlgTagFetcher() {
+    DEBUG_ASSERT(m_pTrackModel);
+    if (!m_pTrackModel) {
+        return;
+    }
+    QModelIndex nextRow = currentTrackInfoIndex.sibling(
+        currentTrackInfoIndex.row()+1, currentTrackInfoIndex.column());
+    if (nextRow.isValid()) {
+        showDlgTagFetcher(nextRow);
+        if (!m_pTrackInfo.isNull()) {
+            showTrackInfo(nextRow);
+        }
+    }
+}
+
+void WTrackMenu::slotPrevDlgTagFetcher() {
+    DEBUG_ASSERT(m_pTrackModel);
+    if (!m_pTrackModel) {
+        return;
+    }
+    QModelIndex prevRow = currentTrackInfoIndex.sibling(
+        currentTrackInfoIndex.row()-1, currentTrackInfoIndex.column());
+    if (prevRow.isValid()) {
+        showDlgTagFetcher(prevRow);
+        if (!m_pTrackInfo.isNull()) {
+            showTrackInfo(prevRow);
+        }
+    }
+}
+
+void WTrackMenu::showDlgTagFetcher(QModelIndex index) {
+    DEBUG_ASSERT(m_pTrackModel);
+    TrackModel* trackModel = m_pTrackModel;
+
+    if (!trackModel) {
+        return;
+    }
+
+    TrackPointer pTrack = trackModel->getTrack(index);
+    currentTrackInfoIndex = index;
+    slotShowTrackInTagFetcher(pTrack);
+}
+
+void WTrackMenu::slotShowTrackInTagFetcher(TrackPointer pTrack) {
+    if (m_pTagFetcher.isNull()) {
+        m_pTagFetcher.reset(new DlgTagFetcher(nullptr));
+        connect(m_pTagFetcher.data(), SIGNAL(next()),
+                this, SLOT(slotNextDlgTagFetcher()));
+        connect(m_pTagFetcher.data(), SIGNAL(previous()),
+                this, SLOT(slotPrevDlgTagFetcher()));
+        connect(m_pTagFetcher.data(), SIGNAL(finished(int)),
+                this, SLOT(slotTagFetcherClosed()));
+    }
+
+    // NULL is fine
+    m_pTagFetcher->loadTrack(pTrack);
+    m_pTagFetcher->show();
+}
```

Comment:
combine these into one function

Reply:
Sure. I was just focussing on decoupling the menu right now. I will try and see if more such improvements can be made.

Reply:
I think you will need to do this to remove the Properties feature's dependence on TrackModel*

Reply:
Okay. Thanks for the pointer (no pun intended). I will refactor most functions for complete decoupling.